provider/apple.go

```diff
@@ -0,0 +1,534 @@
+	}
+	claims := token.Claims{
+		User: &u,
+		StandardClaims: jwt.StandardClaims{
+			Issuer:   ah.Issuer,
+			Id:       cid,
+			Audience: oauthClaims.Audience,
+		},
+		SessionOnly: false,
+	}
+
+	if _, err = ah.JwtService.Set(w, claims); err != nil {
+		rest.SendErrorJSON(w, r, ah.L, http.StatusInternalServerError, err, "failed to set token")
+		return
+	}
+
+	ah.Logf("[DEBUG] user info %+v", u)
+
+	// redirect to back url if presented in login query params
+	if oauthClaims.Handshake != nil && oauthClaims.Handshake.From != "" {
+		http.Redirect(w, r, oauthClaims.Handshake.From, http.StatusTemporaryRedirect)
+		return
+	}
+	rest.RenderJSON(w, &u)
+
+}
+
+// LogoutHandler - GET /logout
+func (ah AppleHandler) LogoutHandler(w http.ResponseWriter, r *http.Request) {
+	if _, _, err := ah.JwtService.Get(r); err != nil {
+		rest.SendErrorJSON(w, r, ah.L, http.StatusForbidden, err, "logout not allowed")
+		return
+	}
+	ah.JwtService.Reset(w)
+}
+
+// exchange sends the validation token request and gets access token and user claims
+// (e.g. https://developer.apple.com/documentation/sign_in_with_apple/generate_and_validate_tokens)
+func (ah *AppleHandler) exchange(ctx context.Context, code, redirectURI string, result interface{}) (err error) {
+
+	// check client_secret is expired and recreate new (client_secret JWT) if need
+	ok, err := ah.isClientSecretExpired()
+	if err != nil || !ok {
+		ah.conf.clientSecret, err = ah.createClientSecret()
+		if err != nil {
+			return errors.Wrap(err, "client secret create failed")
+		}
+	}
+
+	data := url.Values{}
+	data.Set("client_id", ah.conf.ClientID)
+	data.Set("client_secret", ah.conf.clientSecret) // JWT signed with Apple private key
+	data.Set("code", code)
+	data.Set("redirect_uri", redirectURI) // redirect URL can't refer to localhost and must have trusted certificate and https protocol
+	data.Set("grant_type", "authorization_code")
+
+	client := http.Client{Timeout: time.Second * 5}
+	req, err := http.NewRequestWithContext(ctx, "POST", ah.endpoint.TokenURL, strings.NewReader(data.Encode()))
+	if err != nil {
+		return err
+	}
+
+	req.Header.Add("content-type", appleRequestContentType)
+	req.Header.Add("accept", AcceptHeader)
+	req.Header.Add("user-agent", ah.conf.UserAgent) // apple requires a user agent
+
+	res, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+
+	// Apple REST API response either 200 (ok OK) or 400 (if error)
+	if res.StatusCode >= 400 {
+		return fmt.Errorf("exchange response error code: %d", res.StatusCode)
+	}
+
+	err = json.NewDecoder(res.Body).Decode(result)
+	defer func() {
+		if err = res.Body.Close(); err != nil {
+			ah.L.Logf("[ERROR] close request body failed when get access token: %v", err)
+		}
+	}()
+	return err
+}
+
+// getTokenClaims decodes the id_token in response for extract claims with user data
+func (ah AppleHandler) getTokenClaims(idToken string) (jwt.MapClaims, error) {
+	tkn, _, err := new(jwt.Parser).ParseUnverified(idToken, jwt.MapClaims{})
```

Comment:
maybe just pass in (inject) token.jwt and use it instead of manipulating directly? Also, it should have very good reason to parse unverified and I'm not sure what the reason for this

Reply:
I will try parse token claims with `private key` instead unverified parse. If it possible I change implementation for this.